---
output:
  html_document:
    toc: true
    theme: united
  knitrBootstrap::bootstrap_document:
    theme.chooser: TRUE
    highlight.chooser: TRUE
---

Predict coverage
================


TopHat mode: `r tophat`


```{r 'setup'}
library('ballgown')
library('GenomicRanges')
library('usefulstuff')
library('forecast')
library('ggplot2')
library('plyr')
```


# Compute

Loads the coverage and splits it by gene for each sample. Then it calculates the correlation (and $R^2$) between the Geuvadis sample coverage and each of the simulated scenarios including simply drawing numbers from the negative binomial (aka, skipping alignment). For each Geuvadis gene-sample, an ARIMA model is fitted, then one more for each simulated scenario using the simulated coverage as a predictor. The coefficient for the predictor is extracted as well as its corresponding p-value. Similar numbers are calculated between replicates of the same simulation scenario.

```{r 'load'}
## Load coverage data
load(file.path(tophat, 'ex_regCov.Rdata'))

## Define exons
gff <- gffReadGR(file.path('..', 'select_genes', 'twenty_genes.gtf'))
genes <- split(gff, strip_quotes(getAttributeField(as.character(gff$group),     
    'gene_id')))
exons <- unlist(reduce(genes))

## Split by gene
geneCov <- split(ex_regCov, names(exons))
geneCov <- mapply(function(x, y) {
    res <- do.call(rbind, x)
    ## Save the bases
    rownames(res) <- unlist(mapply(seq, start(exons[names(exons) == y]), end(exons[names(exons) == y])))
    return(res)
}, geneCov, names(geneCov), SIMPLIFY = FALSE)


## Split by sample
geu_samples <- c('NA06985', 'NA12144', 'NA12776', 'NA18858', 'NA20542', 
    'NA20772', 'NA20815')


set.seed(20150221)
sampleCov <- lapply(geneCov, function(x) {
    result <- lapply(seq_len(length(geu_samples)), function(i) {
        idx <- grepl(paste0(geu_samples[i], '|s', i), colnames(x))
        res <- x[, idx]
        colnames(res) <- c('Geuvadis', 'd1', 'd2', 'r1', 'r2')
        m <- mean(res$Geuvadis)
        res$b1a <- rnbinom(nrow(res), size = 1, mu = m)
        res$b1b <- rnbinom(nrow(res), size = 1, mu = m)
        res$b6a <- rnbinom(nrow(res), size = 6, mu = m)
        res$b6b <- rnbinom(nrow(res), size = 6, mu = m)
        return(res)
    })
    names(result) <- geu_samples
    return(result)
})


fitArima <- function(x, y = 'Geuvadis', order, d, auto = FALSE) {
    if(auto) {
        f <- auto.arima(d[, y], xreg = d[, x])
        order <- arimaorder(f)
    } else {
        f <- tryCatch(arima(d[, y], xreg = d[, x], order = order), error = function(e) { return(list(error = TRUE))})
    }
    
    if(!is.null(f$error)) return(list(order = NA, coef = NA, z = NA))
    
    z <- coef(f)["d[, x]"] / sqrt(f$var.coef["d[, x]", "d[, x]"])
    return(list(order = paste(order, collapse = '-'), coef = coef(f)["d[, x]"], z = z))
}

rmsd <- function(x, y = 'Geuvadis', d) {
    sqrt(mean( (d[, y] / max(d[, y]) - d[, x] / max(d[, x]))^2 ))
}

corr <- function(x, y = 'Geuvadis', d) { 
    cor(d[, y], d[, x])
}


sets <- c('d1', 'd2', 'r1', 'r2', 'b1a', 'b1b', 'b6a', 'b6b')
reps_a <- sets[rep(c(TRUE, FALSE), 4)]
reps_b <- sets[rep(c(FALSE, TRUE), 4)]

## Save as it takes about 20-30 mins to compute
if(file.exists(file.path(tophat, 'pred_cov.Rdata'))) {
    load(file.path(tophat, 'pred_cov.Rdata'))
} else {
    pred_cov <- lapply(names(sampleCov), function(gene) {
        res <- lapply(names(sampleCov[[gene]]), function(samp) {
            d <- sampleCov[[gene]][[samp]]
        
            ## Arima
            ord <- arimaorder(auto.arima(d$Geuvadis))
            info <- c(lapply(sets, fitArima, order = ord, d = d), mapply(fitArima, x = reps_b, y = reps_a, MoreArgs = list(d = d, order = ord), SIMPLIFY = FALSE))
            orders <- unlist(sapply(info, '[', 'order'))
            coefs <- unlist(sapply(info, '[', 'coef'))
            pvals <- pnorm(abs(unlist(sapply(info, '[', 'z'))), lower = FALSE) * 2
        
            ## RMSD
            RMSD <- c(sapply(sets, rmsd, d = d), mapply(rmsd, reps_b, reps_a, MoreArgs = list(d = d)))
        
            ## Cor and R2
            CORR <- c(sapply(sets, corr, d = d), mapply(corr, reps_b, reps_a, MoreArgs = list(d = d)))
        
        
            data.frame(set = c(sets, paste(reps_a, reps_b, sep = '-')), order = orders, coef = coefs, pval = pvals, rmsd = RMSD, cor = CORR, R2 = CORR^2, type = c(rep(c('d', 'r', 'b1', 'b6'), each = 2), 'd', 'r', 'b1', 'b6'), group = rep(c('Geuvadis', 'replicate'), c(8, 4)), gene = gene, sample = samp, row.names = NULL, stringsAsFactors = FALSE)
        })
        do.call(rbind, res)
    })
    pred_cov <- do.call(rbind, pred_cov)
    pred_cov$set <- factor(pred_cov$set, levels = c(sets, paste(reps_a, reps_b, sep = '-')))

    ## Significant pval?
    pred_cov$sig <- pred_cov$pval <= 0.05

    ## Extract ARIMA order info
    pred_cov$ar <- as.integer(substr(pred_cov$order, 1, 1))
    pred_cov$i <- as.integer(substr(pred_cov$order, 3, 3))
    pred_cov$ma <- as.integer(substr(pred_cov$order, 5, 5))
    save(pred_cov, file = file.path(tophat, 'pred_cov.Rdata'))
}

```

# Models selected

Explore which ARIMA models were selected for each gene-sample scenario.

```{r 'models'}
## Types of models selected
sort(table(pred_cov$order), decreasing = TRUE)
table(pred_cov$ar) /12
table(pred_cov$i) /12
table(pred_cov$ma) /12

ggplot(subset(pred_cov, set == 'd1'), aes(x = ar)) + geom_bar(stat = 'bin') + facet_grid(i ~ ma ) + theme_bw()
```


# Significant coef

Percent of significant ARIMA coefficients for each of the simulated scenarios. Either overall percent, or first summarized by genes or by samples.

```{r 'sig', fig.width = 12}
sig <- ddply(pred_cov, .(set, type), summarize, mean = mean(sig))
ggplot(sig, aes(x = set, y = mean * 100, fill = type)) + geom_boxplot() + ylim(c(0, 100)) + ylab('Percent') + ggtitle('Significant coefs') + theme_bw()
sig <- ddply(pred_cov, .(set, sample, type), summarize, mean = mean(sig))
ggplot(sig, aes(x = set, y = mean * 100, fill = type)) + geom_boxplot(outlier.shape = 5) + ylim(c(0, 100)) + ylab('Percent: 7 samples') + ggtitle('Significant coefs: by genes') + geom_jitter(position = position_jitter(width = 0.2, height = 0), alpha = 1/2) + theme_bw()
sig <- ddply(pred_cov, .(set, gene, type), summarize, mean = mean(sig))
ggplot(sig, aes(x = set, y = mean * 100, fill = type)) + geom_boxplot(outlier.shape = 5) + ylim(c(0, 100)) + ylab('Percent: 20 genes') + ggtitle('Significant coefs: by samples') + geom_jitter(position = position_jitter(width = 0.2, height = 0), alpha = 1/2) + theme_bw()
```


# Pval

Boxplots of ARIMA coefficient p-values. First overall, then by sample (20 points per boxplot: 1 per gene), then by gene (7 points per boxplot).


```{r 'pval', fig.width = 12}
ggplot(pred_cov, aes(x = set, y = pval, fill = type)) + geom_boxplot() + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
ggplot(pred_cov, aes(x = set, y = pval, fill = type)) + geom_boxplot() + facet_grid(. ~ sample) + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
ggplot(pred_cov, aes(x = set, y = pval, fill = type)) + geom_boxplot() + facet_wrap( ~ gene, ncol = 10) + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
```


# Coef

Boxplots of ARIMA coefficients. First overall, then by sample (20 points per boxplot: 1 per gene), then by gene (7 points per boxplot).

```{r 'coef', fig.width = 12}
ggplot(pred_cov, aes(x = set, y = coef, fill = type)) + geom_boxplot() + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
ggplot(pred_cov, aes(x = set, y = coef, fill = type)) + geom_boxplot() + facet_grid(. ~ sample) + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
ggplot(pred_cov, aes(x = set, y = coef, fill = type)) + geom_boxplot() + facet_wrap( ~ gene, ncol = 10) + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
```


# Pval vs Coef

Relationship between p-values and coefficients for the ARIMA models.

```{r 'pval-coef', fig.width = 12}
ggplot(pred_cov, aes(x = pval, y = coef, color = type)) + geom_point() + facet_grid(group ~ .) + geom_vline(xintercept = 0.05, colour = 'red') + theme_bw()
ggplot(pred_cov, aes(x = pval, y = coef, color = type)) + geom_point() + facet_grid(group ~ sample) + geom_vline(xintercept = 0.05, colour = 'red') + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
ggplot(pred_cov, aes(x = pval, y = coef, color = type)) + geom_point() + facet_grid(group ~ gene) + geom_vline(xintercept = 0.05, colour = 'red') + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
```


# RMSD-max

Boxplots of root mean square distance where each vector has been standardized by its maximum. First overall, then by sample (20 points per boxplot: 1 per gene), then by gene (7 points per boxplot).

```{r 'RMSD', fig.width = 12}
ggplot(pred_cov, aes(x = set, y = rmsd, fill = type)) + geom_boxplot() + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
ggplot(pred_cov, aes(x = set, y = rmsd, fill = type)) + geom_boxplot() + facet_grid(. ~ sample) + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
ggplot(pred_cov, aes(x = set, y = rmsd, fill = type)) + geom_boxplot() + facet_wrap( ~ gene, ncol = 10) + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
```


# Correlation

Boxplots of correlation. First overall, then by sample (20 points per boxplot: 1 per gene), then by gene (7 points per boxplot).

```{r 'Correlation', fig.width = 12}
ggplot(pred_cov, aes(x = set, y = cor, fill = type)) + geom_boxplot() + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
ggplot(pred_cov, aes(x = set, y = cor, fill = type)) + geom_boxplot() + facet_grid(. ~ sample) + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
ggplot(pred_cov, aes(x = set, y = cor, fill = type)) + geom_boxplot() + facet_wrap( ~ gene, ncol = 10) + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
```

# Cor vs RMSD-max

Comparison between correlation and RMSD-max.

```{r 'cor-rmsd', fig.width = 12}
ggplot(pred_cov, aes(x = rmsd, y = cor, color = type)) + geom_point() + facet_grid(group ~ .) + theme_bw()
ggplot(pred_cov, aes(x = rmsd, y = cor, color = type)) + geom_point() + facet_grid(group ~ sample) + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
ggplot(pred_cov, aes(x = rmsd, y = cor, color = type)) + geom_point() + facet_grid(group ~ gene) + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
```





# $R^2$

Boxplots of $R^2$ (correlation squared since its a simple linear model). First overall, then by sample (20 points per boxplot: 1 per gene), then by gene (7 points per boxplot).

```{r 'R2', fig.width = 12}
ggplot(pred_cov, aes(x = set, y = R2, fill = type)) + geom_boxplot() + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
ggplot(pred_cov, aes(x = set, y = R2, fill = type)) + geom_boxplot() + facet_grid(. ~ sample) + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8)) 
ggplot(pred_cov, aes(x = set, y = R2, fill = type)) + geom_boxplot() + facet_wrap( ~ gene, ncol = 10) + theme_bw() + theme(axis.text.x  = element_text(angle=90, vjust=0.5, size=8))
```

# Reproducibility

```{r 'reproducibility', dev = 'CairoPNG'}
## Reproducibility info
# Time spent in this report
diff(c(startTime, Sys.time()))
Sys.time()
options(width = 120)
devtools::session_info()
```

